### PR TITLE
Allow by-code scrolling even when by-input scrolling would be denied

### PR DESCRIPTION
**Pull request by davep**
See #1201.

See
https://github.com/Textualize/textual/issues/1201#issuecomment-1380123660 though.

### CODE REVIEW

File: src/textual/widget.py [hunk 1]
```diff
@@ -1853,6 +1902,7 @@ def scroll_visible(
             top (bool, optional): Scroll to top of container. Defaults to False.
             easing (EasingFunction | str | None, optional): An easing method for the scrolling animation. Defaults to "None",
                 which will result in Textual choosing the configured default scrolling easing function.
+            force (bool,optional): Force scrolling even when prohibited by overflow styling. Defaults to `False`.
```
Comment (rodrigogiraoserrao) [inline src/textual/widget.py:1905]:
```suggestion
            force (bool, optional): Force scrolling even when prohibited by overflow styling. Defaults to `False`.
```

File: src/textual/widget.py [hunk 2]
```diff
@@ -1805,6 +1851,7 @@ def scroll_to_region(
             easing (EasingFunction | str | None, optional): An easing method for the scrolling animation. Defaults to "None",
                 which will result in Textual choosing the configured default scrolling easing function.
             top (bool, optional): Scroll region to top of container. Defaults to False.
+            force (bool,optional): Force scrolling even when prohibited by overflow styling. Defaults to `False`.
```
Comment (rodrigogiraoserrao) [inline src/textual/widget.py:1854]:
```suggestion
            force (bool, optional): Force scrolling even when prohibited by overflow styling. Defaults to `False`.
```

File: src/textual/widget.py [hunk 3]
```diff
@@ -1743,6 +1786,7 @@ def scroll_to_widget(
             easing (EasingFunction | str | None, optional): An easing method for the scrolling animation. Defaults to "None",
                 which will result in Textual choosing the configured default scrolling easing function.
             top (bool, optional): Scroll widget to top of container. Defaults to False.
+            force (bool,optional): Force scrolling even when prohibited by overflow styling. Defaults to `False`.
```
Comment (rodrigogiraoserrao) [inline src/textual/widget.py:1789]:
```suggestion
            force (bool, optional): Force scrolling even when prohibited by overflow styling. Defaults to `False`.
```

File: src/textual/widget.py [hunk 4]
```diff
@@ -1708,6 +1748,7 @@ def scroll_page_right(
             duration (float | None, optional): Duration of animation, if animate is True and speed is None.
             easing (EasingFunction | str | None, optional): An easing method for the scrolling animation. Defaults to "None",
                 which will result in Textual choosing the configured default scrolling easing function.
+            force (bool,optional): Force scrolling even when prohibited by overflow styling. Defaults to `False`.
```
Comment (rodrigogiraoserrao) [inline src/textual/widget.py:1751]:
```suggestion
            force (bool, optional): Force scrolling even when prohibited by overflow styling. Defaults to `False`.
```

File: src/textual/widget.py [hunk 5]
```diff
@@ -1677,6 +1714,7 @@ def scroll_page_left(
...
```
Comment (rodrigogiraoserrao) [inline src/textual/widget.py:1717]:
```suggestion
            force (bool, optional): Force scrolling even when prohibited by overflow styling. Defaults to `False`.
```

File: src/textual/widget.py [hunk 6]
```diff
@@ -1648,6 +1682,7 @@ def scroll_page_down(
...
```
Comment (rodrigogiraoserrao) [inline src/textual/widget.py:1685]:
```suggestion
            force (bool, optional): Force scrolling even when prohibited by overflow styling. Defaults to `False`.
```

File: src/textual/widget.py [hunk 7]
```diff
@@ -1590,6 +1618,7 @@ def scroll_up(
...
```
Comment (rodrigogiraoserrao) [inline src/textual/widget.py:1621]:
```suggestion
            force (bool, optional): Force scrolling even when prohibited by overflow styling. Defaults to `False`.
```

File: src/textual/widget.py [hunk 8]
```diff
@@ -1561,6 +1586,7 @@ def scroll_down(
...
```
Comment (rodrigogiraoserrao) [inline src/textual/widget.py:1589]:
```suggestion
            force (bool, optional): Force scrolling even when prohibited by overflow styling. Defaults to `False`.
```

File: src/textual/widget.py [hunk 9]
```diff
@@ -1532,6 +1554,7 @@ def scroll_right(
...
```
Comment (rodrigogiraoserrao) [inline src/textual/widget.py:1557]:
```suggestion
            force (bool, optional): Force scrolling even when prohibited by overflow styling. Defaults to `False`.
```

File: src/textual/widget.py [hunk 10]
```diff
@@ -1503,6 +1522,7 @@ def scroll_left(
...
```
Comment (rodrigogiraoserrao) [inline src/textual/widget.py:1525]:
```suggestion
            force (bool, optional): Force scrolling even when prohibited by overflow styling. Defaults to `False`.
```

File: src/textual/widget.py [hunk 11]
```diff
@@ -1471,6 +1487,7 @@ def scroll_end(
...
```
Comment (rodrigogiraoserrao) [inline src/textual/widget.py:1490]:
```suggestion
            force (bool, optional): Force scrolling even when prohibited by overflow styling. Defaults to `False`.
```

File: src/textual/widget.py [hunk 12]
```diff
@@ -1416,6 +1421,7 @@ def scroll_relative(
...
```
Comment (rodrigogiraoserrao) [inline src/textual/widget.py:1424]:
```suggestion
            force (bool, optional): Force scrolling even when prohibited by overflow styling. Defaults to `False`.
```

File: src/textual/widget.py [hunk 13]
```diff
@@ -1348,10 +1349,13 @@ def scroll_to(
             duration (float | None, optional): Duration of animation, if animate is True and speed is None.
             easing (EasingFunction | str | None, optional): An easing method for the scrolling animation. Defaults to "None",
                 which will result in Textual choosing the default scrolling easing function.
+            force (bool,optional): Force scrolling even when prohibited by overflow styling. Defaults to `False`.
```
Comment (rodrigogiraoserrao) [inline src/textual/widget.py:1352]:
```suggestion
            force (bool, optional): Force scrolling even when prohibited by overflow styling. Defaults to `False`.
```

There are a bunch of these, so maybe Search & Replace and then a single commit? 🤷

File: src/textual/widget.py [hunk 14]
```diff
@@ -803,16 +803,16 @@ def watch_hover_style(
             self.highlight_link_id = hover_style.link_id
 
     def watch_scroll_x(self, old_value: float, new_value: float) -> None:
-        if self.show_horizontal_scrollbar:
+        if self.allow_horizontal_scroll:
             self.horizontal_scrollbar.position = round(new_value)
```
Comment (rodrigogiraoserrao) [inline src/textual/widget.py:807]: I was reading your comment in the original issue and I'm not convinced `allow_horizontal_scroll` should be used here. Note that `scroll_x` was changed, which is why we are in `watch_scroll_x`, so now that `scroll_x` was changed, we need to update the horizontal scrollbar if it is visible.

To me, scrollbar being visible != scrolling enabled, so we really want to check if we are _showing_ the scrollbar.

**Review by rodrigogiraoserrao — CHANGES_REQUESTED** [review]

File: src/textual/widget.py (hunk 14 above)
Comment (willmcgugan) [reply src/textual/widget.py:807]: Hasn't the logic already been done in the various scroll methods, with regards to allowing scrolling?

By the time we get here, I don't think we need to apply the logic again.

This does mean that you can circumvent scroll logic by setting the `scroll_x`  attribute, but I think I'm fine with that.

Comment (davep) [reply src/textual/widget.py:807]: Makes sense to me, I only made that change as it was something you explicitly asked for. I'll nuke that then.

Comment (davep) [reply src/textual/widget.py]: Done.

File: src/textual/widget.py [hunk 15]
```diff
@@ -803,16 +803,12 @@ def watch_hover_style(
             self.highlight_link_id = hover_style.link_id
 
     def watch_scroll_x(self, old_value: float, new_value: float) -> None:
-        if self.show_horizontal_scrollbar:
-            self.horizontal_scrollbar.position = round(new_value)
```
Comment (willmcgugan) [inline src/textual/widget.py:807]: This like is required to change the scrollbar.

Comment (davep) [reply src/textual/widget.py:807]: I'm reading *"like"* as *"line"* there, and assuming it's in reference to updating the scrollbar position. If so, good catch. Restored.

**Pull request merged**